src/Streamly/Internal/Data/Array/Stream/Foreign.hs

```diff
@@ -385,3 +397,140 @@ _toArraysOf :: (MonadIO m, Storable a)
     => Int -> Fold m a (SerialT Identity (Array a))
 _toArraysOf n = FL.chunksOf n (A.writeNF n) FL.toStream
 -}
+
+-------------------------------------------------------------------------------
+-- Parsing
+-------------------------------------------------------------------------------
+
+-- We use a doubly linked list for parseArray. Below is the type (DLL) a few
+-- helper functions (*DLL)
+
+-- Representation of a doubly linked list
+{-# ANN type DLL NoSpecConstr #-}
+newtype DLL a = DLL ([a], a, [a])
+
+{-# INLINE singleDLL #-}
+singleDLL :: a -> DLL a
+singleDLL a = DLL ([], a, [])
+
+{-# INLINE bufferRDLL #-}
+bufferRDLL :: DLL a -> [a]
+bufferRDLL (DLL (_, _, br)) = br
+
+{-# INLINE dropLDLL #-}
+dropLDLL :: DLL a -> DLL a
+dropLDLL (DLL (_, a, x)) = (DLL ([], a, x))
+
+{-# INLINE nextDLL #-}
+nextDLL :: DLL a -> DLL a
+nextDLL (DLL (y, a, x:xs)) = (DLL (a:y, x, xs))
+nextDLL _ = error "nextDLL: Empty right buffer"
+
+{-# INLINE previousDLL #-}
+previousDLL :: DLL a -> DLL a
+previousDLL (DLL (y:ys, a, x)) = (DLL (ys, y, a:x))
+previousDLL _ = error "previousDLL: Empty left buffer"
+
+{-# INLINE currentDLL #-}
+currentDLL :: DLL a -> a
+currentDLL (DLL (_, x, _)) = x
+
+-- Inefficient stuff but probably Ok for our use case.
+{-# INLINE insertDLL #-}
+insertDLL :: a -> DLL a -> DLL a
+insertDLL a (DLL (y, a1, x)) = (DLL (y, a1, x ++ [a]))
+
+{-# INLINE_NORMAL parseD #-}
+parseD ::
+       forall m a b. (MonadIO m, Storable a)
+    => PRD.Parser m a b
+    -> D.Stream m (Array.Array a)
```

Comment:
we should perhaps import Parser/Array types unqualified, becomes a bit easier to read.